### PR TITLE
[SPARK-22404][YARN] Provide an option to use unmanaged AM in yarn-client mode

### PR DESCRIPTION
**Pull request by devaraj-kavali**
## What changes were proposed in this pull request?

Providing a new configuration "spark.yarn.un-managed-am" (defaults to false) to enable the Unmanaged AM Application in Yarn Client mode which launches the Application Master service as part of the Client. It utilizes the existing code for communicating between the Application Master <-> Task Scheduler for the container requests/allocations/launch, and eliminates these,
1. Allocating and launching the Application Master container
2. Remote Node/Process communication between Application Master <-> Task Scheduler

## How was this patch tested?

I verified manually running the applications in yarn-client mode with "spark.yarn.un-managed-am" enabled, and also ensured that there is no impact to the existing execution flows.

I would like to hear others feedback/thoughts on this.

### CODE REVIEW

**Comment by devaraj-kavali** [comment]
@vanzin Thanks for looking into this.

I thought to verify some scenarios before removing WIP, feedback is welcome anytime. Now I see there are some code conflicts, I will resolve conflicts and remove WIP.

**Comment by vanzin** [comment]
ok to test

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 1]
```diff
@@ -69,6 +70,10 @@ private[spark] class Client(
 
   private val isClusterMode = sparkConf.get("spark.submit.deployMode", "client") == "cluster"
 
+  private val isClientUnmanagedAMEnabled =
+    sparkConf.getBoolean("spark.yarn.un-managed-am", false) && !isClusterMode
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: This should be a config constant. Also `unmanagedAM` is more in line with other config names.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 2]
```diff
@@ -656,7 +664,9 @@ private[spark] class Client(
     // Clear the cache-related entries from the configuration to avoid them polluting the
     // UI's environment page. This works for client mode; for cluster mode, this is handled
     // by the AM.
-    CACHE_CONFIGS.foreach(sparkConf.remove)
+    if (!isClientUnmanagedAMEnabled) {
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Why is this needed in the new mode?

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 3]
```diff
@@ -784,6 +794,9 @@ private[spark] class Client(
     val env = new HashMap[String, String]()
     populateClasspath(args, hadoopConf, sparkConf, env, sparkConf.get(DRIVER_CLASS_PATH))
     env("SPARK_YARN_STAGING_DIR") = stagingDirPath.toString
+    if (isClientUnmanagedAMEnabled) {
+      System.setProperty("SPARK_YARN_STAGING_DIR", stagingDirPath.toString)
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Can this be propagated some other way? Using system properties is kinda hacky, and makes it dangerous to run another Spark app later in the same JVM.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 4]
```diff
@@ -1104,14 +1117,39 @@ private[spark] class Client(
...
     }
 
     // Never reached, but keeps compiler happy
     throw new SparkException("While loop is depleted! This should never happen...")
   }
 
+  private def startApplicationMasterService(report: ApplicationReport) = {
+    // Add AMRMToken to establish connection between RM and AM
+    val token = report.getAMRMToken
+    val amRMToken: org.apache.hadoop.security.token.Token[AMRMTokenIdentifier] =
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Why do you need to make this copy? Isn't the `Token` above enough?

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 5]
```diff
@@ -1104,14 +1117,39 @@ private[spark] class Client(
...
+      new org.apache.hadoop.security.token.Token[AMRMTokenIdentifier](token
+        .getIdentifier().array(), token.getPassword().array, new Text(
+        token.getKind()), new Text(token.getService()))
+    val currentUGI = UserGroupInformation.getCurrentUser
+    currentUGI.addToken(amRMToken)
+
+    System.setProperty(
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Same question about using system properties.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 6]
```diff
@@ -51,33 +52,16 @@ import org.apache.spark.util._
 /**
  * Common application master functionality for Spark on Yarn.
  */
-private[spark] class ApplicationMaster(args: ApplicationMasterArguments) extends Logging {
+private[spark] class ApplicationMaster(args: ApplicationMasterArguments, sparkConf: SparkConf,
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: This doesn't follow Spark's convention for multi-line arguments.

This also looks a little odd now, because there are conflicting arguments. `ApplicationMasterArguments` is now only used in cluster mode, and everything else is expected to be provided in the other parameters. So while this is the simpler change, it's also a little ugly.

I don't really have a good suggestion right now, but it's something to think about.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 7]
```diff
@@ -1104,14 +1117,39 @@ private[spark] class Client(
...
+      ApplicationConstants.Environment.CONTAINER_ID.name(),
+      ContainerId.newContainerId(report.getCurrentApplicationAttemptId, 1).toString)
+    val amArgs = new ApplicationMasterArguments(Array("--arg",
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: This is pretty weird, I'd make this an explicit constructor argument for the AM instead. But if I understand this correctly, this is the address the AM will be connecting back to the driver, right? 

It seems like there's an opportunity for better code here, since now they'd both be running in the same process. Like in the cluster mode case, where the AM uses the same `RpcEnv` instance as the driver (see `runDriver()`).

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 8]
```diff
@@ -1104,14 +1117,39 @@ private[spark] class Client(
...
+      sparkConf.get("spark.driver.host") + ":" + sparkConf.get("spark.driver.port")))
+    // Start Application Service in a separate thread and continue with application monitoring
+    new Thread() {
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Don't you want to keep a reference to this thread and join it at some point, to make sure it really goes away? Should it be a daemon thread instead?

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 1 above)
Comment (devaraj-kavali) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Updated the config name and also added config constants.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 2 above)
Comment (devaraj-kavali) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: It is clearing the classpath entries and leading to this error in Executors.
```
Error: Could not find or load main class org.apache.spark.executor.CoarseGrainedExecutorBackend
```

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 4 above)
Comment (devaraj-kavali) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: report.getAMRMToken gives org.apache.hadoop.yarn.api.records.Token type instance, but currentUGI.addToken expects org.apache.hadoop.security.token.Token type instance.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 8 above)
Comment (devaraj-kavali) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: changed it as daemon thread.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 3 above)
Comment (devaraj-kavali) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Changed it to get from the spark conf and the application id.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 5 above)
Comment (devaraj-kavali) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: I changed to set in sparkConf and use the same in ApplicationMaster while getting the containerId.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 7 above)
Comment (devaraj-kavali) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: I added another constructor without `ApplicationMasterArguments` and takes `RpcEnv`  to use the same instance in AM.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala (hunk 6 above)
Comment (devaraj-kavali) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: I made changes to the default constructor and added another constructor. Please check and let me know anything can be done better.

**Comment by devaraj-kavali** [comment]
@vanzin Thanks for the review, can you have a look into the updated PR?

**Comment by vanzin** [comment]
Unlikely I'll be able to review this very soon.

**Comment by HyukjinKwon** [comment]
ok to test

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 2 above)
Comment (vanzin) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: I think this is happening because you're starting the AM after these are removed from the conf. Should probably juggle things around or change how these are provided to the AM, since these configs are super noisy and shouldn't really show up in the UI.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala [hunk 9]
```diff
@@ -183,8 +183,13 @@ object YarnSparkHadoopUtil {
     )
   }
 
-  def getContainerId: ContainerId = {
-    val containerIdString = System.getenv(ApplicationConstants.Environment.CONTAINER_ID.name())
+  def getContainerId(sparkConf: SparkConf): ContainerId = {
+      val containerIdString =
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala]: indentation

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala [hunk 10]
```diff
@@ -183,8 +183,13 @@ object YarnSparkHadoopUtil {
...
+      if (System.getenv(ApplicationConstants.Environment.CONTAINER_ID.name()) != null) {
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala]: better to use `sparkConf.getenv`.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 11]
```diff
@@ -1084,14 +1095,38 @@ private[spark] class Client(
...
+      new org.apache.hadoop.security.token.Token[AMRMTokenIdentifier](token
+        .getIdentifier().array(), token.getPassword().array, new Text(
+        token.getKind()), new Text(token.getService()))
+    val currentUGI = UserGroupInformation.getCurrentUser
+    currentUGI.addToken(amRMToken)
+
+    sparkConf.set("spark.yarn.containerId",
+      ContainerId.newContainerId(report.getCurrentApplicationAttemptId, 1).toString)
+    // Start Application Service in a separate thread and continue with application monitoring
+    val amService = new Thread() {
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Thread name?

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 12]
```diff
@@ -1084,14 +1095,38 @@ private[spark] class Client(
...
 
     // Never reached, but keeps compiler happy
     throw new SparkException("While loop is depleted! This should never happen...")
   }
 
+  private def startApplicationMasterService(report: ApplicationReport) = {
+    // Add AMRMToken to establish connection between RM and AM
+    val token = report.getAMRMToken
+    val amRMToken: org.apache.hadoop.security.token.Token[AMRMTokenIdentifier] =
+      new org.apache.hadoop.security.token.Token[AMRMTokenIdentifier](token
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Keep related calls in the same like (e.g. `token.getIdentifier()`, `new Text(blah)`)

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 13]
```diff
@@ -51,32 +52,27 @@ import org.apache.spark.util._
...
  * Common application master functionality for Spark on Yarn.
  */
-private[spark] class ApplicationMaster(args: ApplicationMasterArguments) extends Logging {
+private[spark] class ApplicationMaster(
+    val args: ApplicationMasterArguments,
+    val sparkConf: SparkConf,
+    val yarnConf: YarnConfiguration)
+  extends Logging {
 
+  def this(sparkConf: SparkConf,
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: See above constructor for multi-line args style.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 14]
```diff
@@ -481,20 +478,29 @@ private[spark] class ApplicationMaster(args: ApplicationMasterArguments) extends
...
-    registerAM(hostname, -1, sparkConf, sparkConf.getOption("spark.driver.appUIAddress"))
-
-    // The driver should be up and listening, so unlike cluster mode, just try to connect to it
-    // with no waiting or retrying.
-    val (driverHost, driverPort) = Utils.parseHostPort(args.userArgs(0))
-    val driverRef = rpcEnv.setupEndpointRef(
-      RpcAddress(driverHost, driverPort),
-      YarnSchedulerBackend.ENDPOINT_NAME)
+    var driverRef : RpcEndpointRef = null
+    if (sparkConf.get(YARN_UNMANAGED_AM)) {
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: I'm not a big fan of this change. Feels like you should have a different method here called `runUnmanaged` that is called instead of `run()`, and takes an `RpcEnv`.

That way you don't need to keep `clientRpcEnv` at all since it would be local to that method, since nothing else here needs it. In fact even `rpcEnv` could go away and become a parameter to `createAllocator`...

**Review by vanzin — COMMENTED** [review]
Hi @devaraj-kavali , sorry this got left behind. It needs updating and I need to think some more about the changes to ApplicationMaster.scala, but it's in the right direction.

**Comment by devaraj-kavali** [comment]
Thanks @vanzin for taking time to look into this, will update it with the changes.

**Comment by vanzin** [comment]
ok to test

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 11 above)
Comment (devaraj-kavali) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: added thread name

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala (hunk 10 above)
Comment (devaraj-kavali) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala]: updated to use sparkConf.getenv

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala (hunk 9 above)
Comment (devaraj-kavali) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala]: corrected the indentation

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala (hunk 13 above)
Comment (devaraj-kavali) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: removed this constructor as part of below comment refactor

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 12 above)
Comment (devaraj-kavali) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: made the change, please let me know if anything better can be done here

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala (hunk 14 above)
Comment (devaraj-kavali) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: Refactored to a method runUnmanaged, please let me know if anything can be done better.

**Comment by devaraj-kavali** [comment]
> since these configs are super noisy and shouldn't really show up in the UI.
These configs are getting removed from sparkConf in ApplicationMaster after using.

**Comment by devaraj-kavali** [comment]
@vanzin can you check the updated changes? thanks

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 15]
```diff
@@ -234,20 +218,26 @@ private[spark] class ApplicationMaster(args: ApplicationMasterArguments) extends
...
+        } else {
+          runExecutorLauncher()
+        }
+      )
     }
     exitCode
   }
 
-  private def runImpl(): Unit = {
+  private def runImpl(opBlock: => Unit): Unit = {
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: There are things in this method that don't look right when you think about an unmanaged AM.

e.g., overriding `spark.master`, `spark.ui.port`, etc, look wrong.

The handling of app attempts also seems wrong, since with an unmanaged AM you don't have multiple attempts. Even the shutdown hooks seems a bit out of place.

Seems to me it would be easier not to try to use this method for the unmanaged AM.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 16]
```diff
@@ -324,6 +311,23 @@ private[spark] class ApplicationMaster(args: ApplicationMasterArguments) extends
...
+        RpcAddress(sparkConf.get("spark.driver.host"),
+          sparkConf.get("spark.driver.port").toInt),
+        YarnSchedulerBackend.ENDPOINT_NAME)
+      // The client-mode AM doesn't listen for incoming connections, so report an invalid port.
+      registerAM(
+        Utils.localHostName, -1, sparkConf, sparkConf.getOption("spark.driver.appUIAddress"))
+      addAmIpFilter(Some(driverRef))
+      createAllocator(driverRef, sparkConf, clientRpcEnv)
+
+      // In client mode the actor will stop the reporter thread.
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: actor?

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 17]
```diff
@@ -617,7 +625,14 @@ private[spark] class ApplicationMaster(args: ApplicationMasterArguments) extends
     try {
       val preserveFiles = sparkConf.get(PRESERVE_STAGING_FILES)
       if (!preserveFiles) {
-        stagingDirPath = new Path(System.getenv("SPARK_YARN_STAGING_DIR"))
+        var stagingDir = System.getenv("SPARK_YARN_STAGING_DIR")
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: `val stagingDir = sys.props.get("...").getOrElse { ... }`

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 18]
```diff
@@ -617,7 +625,14 @@ private[spark] class ApplicationMaster(args: ApplicationMasterArguments) extends
...
+        if (stagingDir == null) {
+          val appStagingBaseDir = sparkConf.get(STAGING_DIR).map { new Path(_) }
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: This looks similar to the logic in `Client.scala`. Maybe the value calculated there should be plumbed through, instead of adding this code.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 19]
```diff
@@ -297,7 +303,10 @@ private[spark] class Client(
             "does not support it", e)
       }
     }
-
+    if (isClientUnmanagedAMEnabled) {
+      // Set Unmanaged AM to true in Application Submission Context
+      appContext.setUnmanagedAM(true)
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: `appContext.setUnmanagedAM(isClientUnmanagedAMEnabled)`

Which also makes the comment unnecessary.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 20]
```diff
@@ -1098,14 +1109,39 @@ private[spark] class Client(
       if (returnOnRunning && state == YarnApplicationState.RUNNING) {
         return createAppReport(report)
       }
-
+      if (state == YarnApplicationState.ACCEPTED && isClientUnmanagedAMEnabled
+        && !amServiceStarted && report.getAMRMToken != null) {
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: indent one more level

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 21]
```diff
@@ -1098,14 +1109,39 @@ private[spark] class Client(
...
+        token.getIdentifier().array(), token.getPassword().array,
+        new Text(token.getKind()), new Text(token.getService()))
+    val currentUGI = UserGroupInformation.getCurrentUser
+    currentUGI.addToken(amRMToken)
+
+    sparkConf.set("spark.yarn.containerId",
+      ContainerId.newContainerId(report.getCurrentApplicationAttemptId, 1).toString)
+    // Start Application Service in a separate thread and continue with application monitoring
+    val amService = new Thread("Unmanaged Application Master Service") {
+      override def run(): Unit = new ApplicationMaster(new ApplicationMasterArguments(Array.empty),
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: This is a pretty long line. Break it down.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 22]
```diff
@@ -1098,14 +1109,39 @@ private[spark] class Client(
...
+    val token = report.getAMRMToken
+    val amRMToken: org.apache.hadoop.security.token.Token[AMRMTokenIdentifier] =
+      new org.apache.hadoop.security.token.Token[AMRMTokenIdentifier](
+        token.getIdentifier().array(), token.getPassword().array,
+        new Text(token.getKind()), new Text(token.getService()))
+    val currentUGI = UserGroupInformation.getCurrentUser
+    currentUGI.addToken(amRMToken)
+
+    sparkConf.set("spark.yarn.containerId",
+      ContainerId.newContainerId(report.getCurrentApplicationAttemptId, 1).toString)
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Won't this name be the same as the first executor created by the app?

I'd rather special-case `getContainerId` to return some baked-in string when the env variable is not set.

Comment (devaraj-kavali) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: made the change to pass the appAttemptId from Client.scala

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 23]
```diff
@@ -324,6 +308,59 @@ private[spark] class ApplicationMaster(args: ApplicationMasterArguments) extends
...
 
+  def runUnmanaged(clientRpcEnv: RpcEnv,
+      appAttemptId: ApplicationAttemptId,
+      stagingDir: Path): Unit = {
+    try {
+      new CallerContext(
+        "APPMASTER", sparkConf.get(APP_CALLER_CONTEXT),
+        Option(appAttemptId.getApplicationId.toString), None).setCurrentContext()
+
+      // This shutdown hook should run *after* the SparkContext is shut down.
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: This is client mode, so you can't rely on shutdown hooks. You need to explicitly stop this service when the SparkContext is shutdown.

Imagine someone just embeds `sc = new SparkContext(); ...; sc.stop()` in their app code, but the app itself runs for way longer than the Spark app.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 24]
```diff
@@ -324,6 +308,59 @@ private[spark] class ApplicationMaster(args: ApplicationMasterArguments) extends
     }
   }
 
+  def runUnmanaged(clientRpcEnv: RpcEnv,
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: Multi-line args start on the next line.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 25]
```diff
@@ -1098,14 +1106,41 @@ private[spark] class Client(
...
+        startApplicationMasterService(report)
+      }
       lastState = state
     }
 
     // Never reached, but keeps compiler happy
     throw new SparkException("While loop is depleted! This should never happen...")
   }
 
+  private def startApplicationMasterService(report: ApplicationReport) = {
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: `: Unit = `

But given you should be explicitly stopping the AM, this should probably return the AM itself.

**Comment by devaraj-kavali** [comment]
@vanzin can you review the latest changes when you have some time? thanks

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 26]
```diff
@@ -744,7 +795,7 @@ private[spark] class ApplicationMaster(args: ApplicationMasterArguments) extends
     override def onDisconnected(remoteAddress: RpcAddress): Unit = {
       // In cluster mode, do not rely on the disassociated event to exit
       // This avoids potentially reporting incorrect exit codes if the driver fails
-      if (!isClusterMode) {
+      if (!(isClusterMode || sparkConf.get(YARN_UNMANAGED_AM))) {
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: Update comment above?

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 27]
```diff
@@ -70,6 +73,11 @@ private[spark] class Client(
 
   private val isClusterMode = sparkConf.get("spark.submit.deployMode", "client") == "cluster"
 
+  private val isClientUnmanagedAMEnabled = sparkConf.get(YARN_UNMANAGED_AM) && !isClusterMode
+  private var amServiceStarted = false
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Do you need this extra flag? Could you just check if `appMaster != null`?

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 28]
```diff
@@ -70,6 +73,11 @@ private[spark] class Client(
...
+  private var appMaster: ApplicationMaster = _
+  private var unManagedAMStagingDirPath: Path = _
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Seems better to just store this in a variable for all cases. It's recomputed from the conf in a few different places in this class.

**Review by vanzin — COMMENTED** [review]
Some minor comments. It would be good to add a test for this in `YarnClusterSuite`.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala [hunk 29]
```diff
@@ -236,6 +236,14 @@ package object config {
       .stringConf
       .createOptional
 
+  /* Unmanaged AM configuration. */
+
+  private[spark] val YARN_UNMANAGED_AM = ConfigBuilder("spark.yarn.unmanagedAM")
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala]: Add `.enabled` to the config key.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 30]
```diff
@@ -298,6 +281,60 @@ private[spark] class ApplicationMaster(args: ApplicationMasterArguments) extends
...
+      createAllocator(driverRef, sparkConf, clientRpcEnv, appAttemptId, cachedResourcesConf)
+      reporterThread.join()
+    } catch {
+      case e: Exception =>
+        // catch everything else if not specifically handled
+        logError("Uncaught exception: ", e)
+        finish(FinalApplicationStatus.FAILED,
+          ApplicationMaster.EXIT_UNCAUGHT_EXCEPTION,
+          "Uncaught exception: " + StringUtils.stringifyException(e))
+        if (!unregistered) {
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: Is this code needed here? Won't it be called when the client calls `stopUnmanaged`?

File: resource-managers/yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnClusterSuite.scala [hunk 31]
```diff
@@ -87,6 +87,10 @@ class YarnClusterSuite extends BaseYarnClusterSuite {
     testBasicYarnApp(false)
   }
 
+  test("run Spark in yarn-client mode with unmanaged am") {
+    testBasicYarnApp(true, Map("spark.yarn.unmanagedAM.enabled" -> "true"))
```
Comment (vanzin) [inline resource-managers/yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnClusterSuite.scala]: `YARN_UNMANAGED_AM.key`

**Review by vanzin — COMMENTED** [review]
What's the behavior when the AM fails before the context is stopped?

From the code I see some stuff is printed to the logs and the YARN app is marked as finished. But does the context remain alive? Or should that event cause the context to be stopped?

I'm mostly concerned with how clear it is for the user that the jobs start failing because the context is now unusable.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala (hunk 30 above)
Comment (devaraj-kavali) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: `appMaster.runUnmanaged` is running in a daemon thread, if something goes unexpected in `appMaster.runUnmanaged` then the daemon thread stops and monitor thread will not know about it and continue with the status as ACCEPTED/RUNNING, this code unregisters with RM so that the Client/monitor thread gets the application report status as FAILED and stops the services including `sc`.

**Comment by devaraj-kavali** [comment]
> What's the behavior when the AM fails before the context is stopped?
> 
> From the code I see some stuff is printed to the logs and the YARN app is marked as finished. But does the context remain alive? Or should that event cause the context to be stopped?
> 
> I'm mostly concerned with how clear it is for the user that the jobs start failing because the context is now unusable.

If AM fails before the context is stopped, AM reports the FAILED status to the RM and Client receives the FAILED status as part of monitoring and stops the services including the context.

**Comment by vanzin** [comment]
> If AM fails before the context is stopped...

That explains the YARN side of things. What about the Spark side of things? What will the user see? Is it clear to the user that the context is now unusable because the YARN app is not running anymore?

Or will they get weird errors because of executors not being allocated and things like that?

**Comment by devaraj-kavali** [comment]
> That explains the YARN side of things. What about the Spark side of things? What will the user see? Is it clear to the user that the context is now unusable because the YARN app is not running anymore?
> 
> Or will they get weird errors because of executors not being allocated and things like that?

It gives an error log from YarnClientSchedulerBackend that `YARN application has exited unexpectedly with state FAILED!...` and shows the user that `(No active SparkContext.)` when try to access, and no additional errors, the behavior is similar to application failed in yarn-client mode without unmanaged am enabled.

**Comment by vanzin** [comment]
Sounds good. Merging to master.

**Comment by devaraj-kavali** [comment]
Thank you so much @vanzin.
